src/isort/helpers.rs

```diff
@@ -1,7 +1,37 @@
 use rustpython_ast::Stmt;
+use rustpython_parser::lexer;
+use rustpython_parser::lexer::Tok;
 
+use crate::ast::types::Range;
+use crate::isort::types::TrailingComma;
 use crate::source_code_locator::SourceCodeLocator;
 
+/// Return `true` if a `StmtKind::ImportFrom` statement ends with a magic
+/// trailing comma.
+pub fn trailing_comma(stmt: &Stmt, locator: &SourceCodeLocator) -> TrailingComma {
+    let contents = locator.slice_source_code_range(&Range::from_located(stmt));
+    let mut count: usize = 0;
+    let mut trailing_comma = TrailingComma::Absent;
+    for (_, tok, _) in lexer::make_tokenizer(&contents).flatten() {
+        if matches!(tok, Tok::Lpar) {
+            count += 1;
+        }
+        if matches!(tok, Tok::Rpar) {
+            count -= 1;
+        }
+        if count == 1 {
+            if matches!(tok, Tok::Newline | Tok::Indent | Tok::Dedent | Tok::Comment) {
+                continue;
+            } else if matches!(tok, Tok::Comma) {
+                trailing_comma = TrailingComma::Present;
+            } else {
+                trailing_comma = TrailingComma::Absent;
+            }
+        }
+    }
+    trailing_comma
+}
```

Comment:
@colin99d - I went with a slightly different approach here. The previous version relied on passing around the entire source code and iterating over characters, which ran two risks: (1) it's kind of inefficient to grab the entire source code; and (2) the logic (while working!) required handling a variety of cases, since we were matching on raw strings.

This version tokenizes the code, so that we can just look for the `Tok::Comma` as the "last" token in the import (excluding some unimportant tokens, like newlines).

Reply:
This is much better! I was assuming there were better ways to do things that would come out in the review.